### PR TITLE
[SYCLomatic] Allow non-standard syntax for template method out-of-class definition for CUDA language and fix it in the migrated code

### PR DESCRIPTION
**Pull request by zhiweij1**
Signed-off-by: Jiang, Zhiwei <zhiwei.jiang@intel.com>

### CODE REVIEW

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 1]
```diff
@@ -14613,3 +14613,26 @@ void TypeRemoveRule::runRule(
 }
 
 REGISTER_RULE(TypeRemoveRule, PassKind::PK_Analysis)
+
+void CompatNVCCRule::registerMatcher(ast_matchers::MatchFinder &MF) {
+  MF.addMatcher(functionDecl().bind("FunctionDecl"), this);
```
Comment (ShengchenJ) [inline clang/lib/DPCT/ASTTraversal.cpp]: How about use the functionTemplateDecl as a replace of functionDecl?

Comment (zhiweij1) [reply clang/lib/DPCT/ASTTraversal.cpp]: I have updated the matcher.

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 2]
```diff
@@ -14613,3 +14613,28 @@ void TypeRemoveRule::runRule(
 }
 
 REGISTER_RULE(TypeRemoveRule, PassKind::PK_Analysis)
+
+void CompatNVCCRule::registerMatcher(ast_matchers::MatchFinder &MF) {
```
Comment (zhimingwang36) [inline clang/lib/DPCT/ASTTraversal.cpp]: how about change the rule to: CompatWithClang,  add some comment to detail on diff b/w EDG front end and Clang.

File: clang/lib/Sema/SemaDecl.cpp [hunk 3]
```diff
@@ -10540,8 +10540,20 @@ Sema::ActOnFunctionDeclarator(Scope *S, Declarator &D, DeclContext *DC,
         HasExplicitTemplateArgs = false;
       } else if (FunctionTemplate) {
         // Function template with explicit template arguments.
+#ifdef SYCLomatic_CUSTOMIZATION
+
+        if (getLangOpts().CUDA) {
```
Comment (zhimingwang36) [inline clang/lib/Sema/SemaDecl.cpp:10544]: Please check if need add this condition.

Comment (zhiweij1) [reply clang/lib/Sema/SemaDecl.cpp:10544]: `nvcc -x c++` cannot compile this code also. So it is reasonable to add this condition.

File: clang/lib/DPCT/ASTTraversal.cpp (hunk 2 above)
Comment (zhiweij1) [reply clang/lib/DPCT/ASTTraversal.cpp]: Done

**Review by tomflinda — APPROVED** [review]
LGTM

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**